File: Sprint7/G/C++/main.cpp
// https://contest.yandex.ru/contest/25596/problems/G/
#include <iostream>
#include <vector>

int main() {
    int m, n;
    std::cin >> m >> n;

    std::vector<int> coins(n);
    for(int i = 0; i < n; i++) {
        std::cin >> coins[i];
    }

    std::vector<long long> dp(m + 1, 0);
    dp[0] = 1;

    for(int d : coins) {
        for(int i = d; i <= m; i++) {
            dp[i] += dp[i - d];
        }
    }

    std::cout << dp[m] << std::endl;

    return 0;
}

File: Sprint7/G/Rust/src/main.rs
// https://contest.yandex.ru/contest/25596/problems/G/

fn main() {
    let mut lines = std::io::stdin().lines();
    let m = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
    let n = lines.next().unwrap().unwrap().parse::<usize>().unwrap();

    let coins = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap())
        .collect::<Vec<usize>>();

    let mut dp = vec![0; m + 1];
    dp[0] = 1;

    for coin in coins {
        for i in coin..(m + 1) {
            dp[i] += dp[i - coin];
        }
    }

    println!("{}", dp[m]);
}
